### ISSUE TITLE
Model as close to OpenNMT

### ISSUE THREAD

**Issue by RikVN** [marian-nmt/marian#222]
I've been looking into switching from OpenNMT to Marian since it is a lot faster. However, I've been unable to reproduce the results I got with OpenNMT. I do semantic parsing, which means small data sets (4-5k gold examples), but with OpenNMT I still got (very) good results. To give a bit of insight, I used to score 0.75-0.80 (F-score), but now can't get higher than 0.50.

These are the OpenNMT settings I use (training sets etc omitted):

th train.lua -encoder_type brnn  -src_seq_length 1000 -tgt_seq_length 1000 -src_word_vec_size 300 -tgt_word_vec_size 300 -layers 2 -rnn_size 300 -rnn_type LSTM -dropout 0.2  -dropout_words 0 -dropout_type naive -bridge copy -attention global -max_pos 1000 -global_attention general -validation_metric perplexity -max_batch_size 12 -optim sgd -learning_rate 0.7 -max_grad_norm 5 -learning_rate_decay 0.7 -start_decay_at 9
 
For Marian, I used these settings as a starting point (also train sets etc omitted):

./marian --type s2s --dim-emb 300 --dim-rnn 300 --enc-type bidirectional --enc-cell lstm --enc-depth 2 --dec-cell lstm --dec-depth 2 --dropout-rnn 0.2 --max-length 1000  -e 12 --disp-freq 350 --mini-batch 12 -o sgd -l 0.7 --lr-decay 0.7 --lr-decay-strategy epoch --lr-decay-start 9 --clip-norm 1 --valid-metrics perplexity --valid-mini-batch 12 --valid-max-length 1000 - -b 10 -n 0.9 --max-length-factor 100 --cost-type perplexity

But that only scores 0.35, as compared to my previous scores of 0.75-0.80.

Starting from these settings, I did experiment varying these parameters:

--type {amun, nematus} - no improvement
--dim {256, 512, 1024} - no improvement
--depth {1} - no improvement
--cell {gru, tanh} - no improvement
--optimizer {adam, adagrad} - adam scores a bit better than sgd
--cost-type {cross-entropy} - scores better than perplexity (0.1 increase to 0.50)
--batch_size {6, 24, 64} - no improvement

The experiments go quite fast, so I can do more parameter searches, but I'm wondering: am I missing a certain setting that's very important? Is there something that OpenNMT does that's not implemented here? Are there inherent settings in Marian that do not work well with small data sets?

Thanks in advance for answering!

**Comment by emjotde** [comment]
Hi, 
Can you try this? I purposefully removed all learning-rate and optimizer tricks. OpenNMT has a strong believe in plain SGD, while we never really use SGD at all, so there might be issues with our code in that respect (barely tested). 
On the other hand default values for ADAM usually work really well with any kind of problem we ever tried. Maybe let’s start here? If that improves things I will be happy to help you explore a couple of learning rate schemes.
./marian --type s2s --dim-emb 300 --dim-rnn 300 --enc-type bidirectional --enc-cell lstm --enc-depth 2 --dec-cell lstm --dec-depth 2 --dropout-rnn 0.2 --max-length 1000 -e 12 --disp-freq 350 --mini-batch 2 --valid-metrics perplexity --valid-mini-batch 12 --valid-max-length 1000 - -b 10 -n 0.9 --max-length-factor 100 --cost-type ce-mean-words
Other than that, how large is your data set? There may be far better settings in terms of mini-batch size etc.


From: Rik van Noord
Sent: Tuesday, November 13, 2018 11:47 AM
To: marian-nmt/marian
Cc: Subscribed
Subject: [marian-nmt/marian] Model as close to OpenNMT (#222)

I've been looking into switching from OpenNMT to Marian since it is a lot faster. However, I've been unable to reproduce the results I got with OpenNMT. I do semantic parsing, which means small data sets (4-5k gold examples), but with OpenNMT I still got (very) good results. To give a bit of insight, I used to score 0.75-0.80 (F-score), but now can't get higher than 0.50.
These are the OpenNMT settings I use (training sets etc omitted):
th train.lua -encoder_type brnn -src_seq_length 1000 -tgt_seq_length 1000 -src_word_vec_size 300 -tgt_word_vec_size 300 -layers 2 -rnn_size 300 -rnn_type LSTM -dropout 0.2 -dropout_words 0 -dropout_type naive -bridge copy -attention global -max_pos 1000 -global_attention general -validation_metric perplexity -max_batch_size 12 -optim sgd -learning_rate 0.7 -max_grad_norm 5 -learning_rate_decay 0.7 -start_decay_at 9
For Marian, I used these settings as a starting point (also train sets etc omitted):
./marian --type s2s --dim-emb 300 --dim-rnn 300 --enc-type bidirectional --enc-cell lstm --enc-depth 2 --dec-cell lstm --dec-depth 2 --dropout-rnn 0.2 --max-length 1000 -e 12 --disp-freq 350 --mini-batch 12 -o sgd -l 0.7 --lr-decay 0.7 --lr-decay-strategy epoch --lr-decay-start 9 --clip-norm 1 --valid-metrics perplexity --valid-mini-batch 12 --valid-max-length 1000 - -b 10 -n 0.9 --max-length-factor 100 --cost-type perplexity
But that only scores 0.35, as compared to my previous scores of 0.75-0.80.
Starting from these settings, I did experiment varying these parameters:
--type {amun, nematus} - no improvement
--dim {256, 512, 1024} - no improvement
--depth {1} - no improvement
--cell {gru, tanh} - no improvement
--optimizer {adam, adagrad} - adam scores a bit better than sgd
--cost-type {cross-entropy} - scores better than perplexity (0.1 increase to 0.50)
--batch_size {6, 24, 64} - no improvement
The experiments go quite fast, so I can do more parameter searches, but I'm wondering: am I missing a certain setting that's very important? Is there something that OpenNMT does that's not implemented here? Are there inherent settings in Marian that do not work well with small data sets?
Thanks in advance for answering!

**Comment by emjotde** [comment]
Hi, 
Can you try this? I purposefully removed all learning-rate and optimizer tricks. OpenNMT has a strong believe in plain SGD, while we never really use SGD at all, so there might be issues with our code in that respect (barely tested). 

On the other hand default values for ADAM usually work really well with any kind of problem we ever tried. Maybe let’s start here? If that improves things I will be happy to help you explore a couple of learning rate schemes.

```
./marian --type s2s --dim-emb 300 --dim-rnn 300 \
 --enc-type bidirectional --enc-cell lstm --enc-depth 2 --dec-cell lstm --dec-depth 2 \
 --dropout-rnn 0.2 --max-length 1000 -e 12 --disp-freq 350 --mini-batch 12 \
 --valid-metrics perplexity --valid-mini-batch 12 --valid-max-length 1000 \
 -b 10 -n 0.9 --max-length-factor 100 --cost-type ce-mean-words
```

Other than that, how large is your data set? There may be far better settings in terms of mini-batch size etc.

**Comment by emjotde** [comment]
I wonder what dropout 0.2 means in OpenNMT. I think with that few instances we might be overfitting everywhere else too. `dropout-rnn` only regularizes the RNN. We do not have other regularization setting as this was never really necessary. Maybe try to set `dropout-rnn` a bit higher, and experiment with `dropout-src` and maybe `label-smoothing 0.1`. If that helps we have a hint and I can quickly add regularization to embeddings and output layer.

**Comment by emjotde** [comment]
To be honest, I don't have any answers for you. Your tiny dataset is beyond anything I ever tried.
